File: python/tests/test_config.py
from pathlib import Path

import pytest

from vision_pipeline.cli import main
from vision_pipeline.config import (
    PipelineConfig,
    SourceKind,
    parse_source,
    positive_integer,
    unit_interval,
)


@pytest.mark.parametrize("value, expected", [("0", 0), ("12", 12), ("001", 1)])
def test_camera_index(value: str, expected: int) -> None:
    source = parse_source(value)
    assert source.kind is SourceKind.CAMERA
    assert source.location == expected


@pytest.mark.parametrize("value", ["", " ", "-1", "+1", "1.5", "2147483648"])
def test_invalid_camera_index(value: str) -> None:
    with pytest.raises(ValueError, match="camera|Camera"):
        parse_source(value)


@pytest.mark.parametrize("suffix, kind", [(".PNG", SourceKind.IMAGE), (".mp4", SourceKind.VIDEO)])
def test_file_source(tmp_path: Path, suffix: str, kind: SourceKind) -> None:
    path = tmp_path / f"source{suffix}"
    path.touch()  # Parsing checks paths; decoding is acquisition's responsibility.
    assert parse_source(str(path)).kind is kind
    assert parse_source(str(path)).location == path


def test_source_errors(tmp_path: Path) -> None:
    with pytest.raises(ValueError, match="does not exist"):
        parse_source(str(tmp_path / "missing.png"))
    with pytest.raises(ValueError, match="not a regular file"):
        parse_source(str(tmp_path))
    path = tmp_path / "data.txt"
    path.touch()
    with pytest.raises(ValueError, match="Unsupported source extension"):
        parse_source(str(path))
    with pytest.raises(ValueError, match="Unsupported source URL"):
        parse_source("https://example.test/video.mp4")


@pytest.mark.parametrize("value", ["abc", "nan", "inf", "-inf", "-0.01", "1.01", "", True])
def test_invalid_threshold(value) -> None:
    with pytest.raises(ValueError, match="between 0 and 1"):
        unit_interval(value, "--confidence")


@pytest.mark.parametrize("value", ["0", "1", "0.5", "1e-2"])
def test_valid_threshold(value: str) -> None:
    assert unit_interval(value, "--iou") == float(value)


@pytest.mark.parametrize("value", ["0", "-1", "1.5", "abc", "", "inf"])
def test_invalid_frame_limit(value: str) -> None:
    with pytest.raises(ValueError, match="positive integer"):
        positive_integer(value)


def test_frame_limit() -> None:
    assert positive_integer("10") == 10


@pytest.mark.parametrize("value", [0, -1, 1.5, True, "2"])
def test_dataclass_validates_frame_limit(value) -> None:
    with pytest.raises(ValueError, match="positive integer"):
        PipelineConfig(parse_source("0"), max_frames=value)


@pytest.mark.parametrize("field", ["confidence", "iou"])
def test_dataclass_validates_threshold(field: str) -> None:
    with pytest.raises(ValueError, match=field):
        PipelineConfig(parse_source("0"), **{field: float("nan")})


def test_output_validation(image_path: Path, tmp_path: Path) -> None:
    source = parse_source(str(image_path))
    with pytest.raises(ValueError, match="differ from the source"):
        PipelineConfig(source, output=image_path)
    alias = tmp_path / "alias.png"
    alias.symlink_to(image_path)
    with pytest.raises(ValueError, match="differ from the source"):
        PipelineConfig(source, output=alias)
    alias.unlink()
    alias.hardlink_to(image_path)
    with pytest.raises(ValueError, match="differ from the source"):
        PipelineConfig(source, output=alias)
    with pytest.raises(ValueError, match="Unsupported output extension"):
        PipelineConfig(source, output=tmp_path / "out.mp4")
    with pytest.raises(ValueError, match="Unsupported output extension"):
        PipelineConfig(parse_source("0"), output=tmp_path / "out.png")
    directory = tmp_path / "directory.png"
    directory.mkdir()
    with pytest.raises(ValueError, match="not a regular file"):
        PipelineConfig(source, output=directory)
    with pytest.raises(ValueError, match="parent is not a directory"):
        PipelineConfig(source, output=image_path / "out.png")
    output = tmp_path / "new" / "out.png"
    assert PipelineConfig(source, output=output).output == output
    assert not output.parent.exists()  # Validation has no write side effects.


@pytest.mark.parametrize(
    "arguments, message",
    [
        (["--no-display"], "--source is required"),
        (["--source", "0", "--confidence", "bad"], "--confidence"),
        (["--source", "0", "--iou", "nan"], "--iou"),
        (["--source", "0", "--max-frames", "0"], "--max-frames"),
        (["--source", "0", "--output", ""], "--output must be a non-empty"),
    ],
)
def test_cli_configuration_errors(arguments, message: str, capsys) -> None:
    with pytest.raises(SystemExit) as error:
        main(arguments)
    assert error.value.code == 2
    assert message in capsys.readouterr().err

File: cpp/tests/test_config.cpp
#include "test_support.hpp"
#include "vision/config.hpp"

#include <fstream>
#include <vector>

namespace {

vision::CliOptions parse(std::vector<std::string> arguments) {
  std::vector<char*> argv;
  argv.reserve(arguments.size());
  for (auto& argument : arguments) argv.push_back(argument.data());
  return vision::parse_cli(static_cast<int>(argv.size()), argv.data());
}

}  // namespace

void run_config_tests() {
  using test_support::check;
  using test_support::expect_throw;
  using test_support::TemporaryDirectory;

  TemporaryDirectory directory;
  const auto image = directory.path() / "fixture.PNG";
  std::ofstream(image) << "fixture";
  const auto video = directory.path() / "fixture.avi";
  std::ofstream(video) << "fixture";
  const auto model = directory.path() / "model.onnx";
  std::ofstream(model) << "fixture";
  const auto labels = directory.path() / "labels.txt";
  {
    std::ofstream label_stream(labels);
    for (int index = 0; index < 80; ++index) label_stream << "class" << index << '\n';
  }

  const auto camera = vision::parse_source("12");
  check(camera.kind == vision::SourceKind::camera && camera.camera_index == 12, "Camera parsing failed.");
  check(vision::parse_source(image.string()).kind == vision::SourceKind::image, "Image parsing failed.");
  check(vision::parse_source(video.string()).kind == vision::SourceKind::video, "Video parsing failed.");
  expect_throw([] { vision::parse_source("-1"); }, "non-negative");
  expect_throw([] { vision::parse_source("missing.png"); }, "does not exist");
  expect_throw([] { vision::parse_source("https://example.test/a.mp4"); }, "Unsupported source URL");
  expect_throw([] { vision::parse_unit_interval("1.1", "--confidence"); }, "between 0 and 1");
  expect_throw([] { vision::parse_unit_interval("nan", "--iou"); }, "between 0 and 1");
  expect_throw([] { vision::parse_positive_integer("0", "--max-frames"); }, "positive integer");
  expect_throw([] { vision::parse_nonnegative_integer("-1", "--warmup"); }, "non-negative");

  const auto options = parse({"vision_cpp", "--source", image.string(), "--output",
                              (directory.path() / "result.jpg").string(), "--no-display", "--max-frames", "1",
                              "--confidence", "0", "--iou", "1"});
  check(options.action == vision::CliAction::run && options.config->no_display, "CLI parsing failed.");
  check(options.config->max_frames == 1U, "Maximum frame parsing failed.");
  check(parse({"vision_cpp"}).action == vision::CliAction::help, "No-argument help failed.");
  check(parse({"vision_cpp", "--no-display"}).action == vision::CliAction::help, "Missing-source help failed.");
  check(parse({"vision_cpp", "--version"}).action == vision::CliAction::version, "Version parsing failed.");
  expect_throw([&] { parse({"vision_cpp", "--source", image.string(), "--output", "result.txt"}); },
               "Unsupported output extension");
  expect_throw([&] { parse({"vision_cpp", "--source", image.string(), "--output", image.string()}); },
               "Output must differ");
  expect_throw([&] { parse({"vision_cpp", "--source", image.string(), "--detections-json", "result.json"}); },
               "requires --model");
  expect_throw([&] { parse({"vision_cpp", "--source", image.string(), "--model", model.string(), "--labels", labels.string(), "--detections-json", "result.txt"}); },
               "must be a .json");
  expect_throw([&] { parse({"vision_cpp", "--source", video.string(), "--model", model.string(), "--labels", labels.string(), "--detections-json", "result.json"}); },
               "requires an image source");
  expect_throw([&] { parse({"vision_cpp", "--source", video.string(), "--benchmark"}); },
               "requires --no-display");
  expect_throw([&] { parse({"vision_cpp", "--source", video.string(), "--no-display", "--benchmark-output", "result.json"}); },
               "requires --benchmark");
  expect_throw([&] { parse({"vision_cpp", "--source", video.string(), "--no-display", "--benchmark", "--benchmark-output", "result.txt"}); },
               "must be a .json");
}
